### PR TITLE
Ana Maria Dobrescu HW02

### PR DESCRIPTION
**Pull request by dobreasca**

### CODE REVIEW

File: Main.hs [hunk 1]
```diff
@@ -11,144 +11,156 @@ import qualified Data.List as L
...
 
-dotProduct :: [Int] -> [Int] -> Int 
-dotProduct = undefined
+dotProduct :: [Int] -> [Int] -> Int
+dotProduct x y = sum (zipWith (*) x y)
 
 applyAll :: [a -> b] -> a -> [b]
-applyAll = undefined
+applyAll [] _ = []
+applyAll (f : fs) a = f a : applyAll fs a
```
Comment (alex28sh) [inline Main.hs:42]: you can use `map` function

File: Main.hs [hunk 2]
```diff
@@ -11,144 +11,156 @@ import qualified Data.List as L
...
 
-interleave :: [a] -> [a] -> [a] 
-interleave = undefined
+interleave :: [a] -> [a] -> [a]
+interleave [] _ = []
+interleave _ [] = []
+interleave (x : xs) (y : ys) = x : y : interleave xs ys
```
Comment (alex28sh) [inline Main.hs:47]: too many code, use higher-order functions (like `zipWith`)

File: Main.hs [hunk 3]
```diff
@@ -11,144 +11,156 @@ import qualified Data.List as L
 import Text.Printf (printf)
 
 multByIndex :: [Int] -> [Int]
-multByIndex = undefined
+multByIndex xs = [x * i | (x, i) <- zip xs [0 ..]]
 
 powerByIndex :: [Int] -> [Int]
-powerByIndex = undefined
+powerByIndex xs = [x ^ i | (x, i) <- zip xs [0 ..]]
```
Comment (alex28sh) [inline Main.hs:17]: same)

**Review by alex28sh — CHANGES_REQUESTED** [review]

File: Main.hs (hunk 3 above)
Comment (dobreasca) [reply Main.hs]: I did the changes you requested in main.hs. is it ok now?

File: Main.hs [hunk 4]
```diff
@@ -11,40 +11,31 @@ import qualified Data.List as L
...
-isSorted [_] = True
-isSorted (x : y : xs) = if y < x then False else isSorted (y : xs)
+isSorted xs = all (uncurry (<=)) (zip xs (tail xs))
 
 countElement :: Int -> [Int] -> Int
-countElement _ [] = 0
-countElement a (x : xs) = cnt + countElement a xs
-  where
-    cnt = if a == x then 1 else 0
+countElement a = foldl (\cnt x -> if x == a then cnt + 1 else cnt) 0
```
Comment (kajigor) [inline Main.hs:29]: This is a good solution; however, there is a more idiomatic one, which includes filtering. Can you implement it?

Comment (dobreasca) [reply Main.hs]: changed it :)

**Comment by kajigor** [comment]
Good 👍

**Pull request closed without merging**